src/disk_cache.rs

```diff
@@ -0,0 +1,956 @@
+            size: meta.size,
+            e_tag: meta.e_tag,
+            version: meta.version,
+        }))
+    }
+
+    // only used in test
+    pub async fn cached_parts(&self) -> object_store::Result<Vec<PartID>> {
+        let pattern = self.make_part_path(0, true);
+        let mut part_paths = glob::glob(&pattern.to_string_lossy())
+            .map_err(wrap_io_err)?
+            .filter_map(Result::ok)
+            .collect::<Vec<_>>();
+
+        // not cached at all
+        if part_paths.is_empty() {
+            return Ok(vec![]);
+        }
+
+        // sort the paths in alphabetical order
+        part_paths.sort();
+
+        // retrieve the part numbers from the paths
+        let mut part_numbers = Vec::with_capacity(part_paths.len());
+        for part_path in part_paths.iter() {
+            let file_ext = match part_path.extension() {
+                None => continue,
+                Some(ext) => ext.to_string_lossy(),
+            };
+            let part_number = file_ext
+                .split('-')
+                .last()
+                .and_then(|part_number| part_number.parse::<usize>().ok());
+            if let Some(part_number) = part_number {
+                part_numbers.push(part_number);
+            }
+        }
+
+        Ok(part_numbers)
+    }
+
+    // if the disk is full, we'll get an error here.
+    // TODO: this error can be ignored in the caller side and print a warning.
+    async fn save_part(&self, part_number: usize, buf: &[u8]) -> object_store::Result<()> {
+        let part_file_path = self.make_part_path(part_number, false);
+
+        // if the part already exists, do not save again.
+        if Some(true) == fs::try_exists(&part_file_path).await.ok() {
+            return Ok(());
+        }
+
+        // ensure the parent folder exists
+        if let Some(part_file_folder) = part_file_path.parent() {
+            fs::create_dir_all(part_file_folder)
+                .await
+                .map_err(wrap_io_err)?;
+        }
+
+        // save the file content to tmp. if the disk is full, we'll get an error here.
+        let tmp_part_file_path =
+            part_file_path.with_extension(format!("tmp{}", self.make_rand_suffix()));
+        let mut file = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(&tmp_part_file_path)
+            .await
+            .map_err(wrap_io_err)?;
+        file.write_all(buf).await.map_err(wrap_io_err)?;
+
+        // rename the file
+        fs::rename(tmp_part_file_path, part_file_path)
+            .await
+            .map_err(wrap_io_err)
+            .ok();
+        Ok(())
+    }
+
+    async fn save_meta(&self, meta: &ObjectMeta) -> object_store::Result<()> {
+        let meta_file_path = self.make_meta_path();
+
+        // if the meta file exists and not corrupted, do nothing
+        match self.read_meta().await {
+            Ok(Some(_)) => return Ok(()),
```

Comment:
I guess this is safe to do for us since our files are all immutable. 👍

Reply:
yes, making all the files immutable makes caching much easier :)